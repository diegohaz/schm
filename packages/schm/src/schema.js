// @flow
import merge from 'lodash/merge'
import type { Schema, SchemaGroup } from './types'
import { isSchema } from './utils'
import parsers from './parsers'
import validators from './validators'
import parse from './parse'
import validate from './validate'
import parseParams from './parseParams'

export const defaultSchema = (params?: Object = {}): Schema => ({
  parsers,
  validators,
  parse(values) {
    return parse(values, this)
  },
  validate(values, paramPathPrefix) {
    return validate(values, this, paramPathPrefix)
  },
  merge(...schemas) {
    const merged = merge({}, this, ...schemas)
    return {
      ...merged,
      params: parseParams(merged.params),
    }
  },
  params: parseParams(params),
})

/**
 * A simple group of parameters. It's used internally when you pass literal
 * objects to [`schema`](#schema).
 * @example
 * const userSchema = schema(
 *   group({
 *     id: String,
 *     name: String,
 *   }),
 *   group({
 *     age: Number,
 *   })
 * )
 */
export const group = (params?: Object = {}): SchemaGroup => (
  (previous: Schema): Schema => previous.merge(defaultSchema(params))
)

/**
 * Creates a schema by composing groups of parameters.
 * @example
 * const userSchema = schema({
 *   id: String,
 *   name: String,
 * }, {
 *   age: Number
 * })
 *
 * // nested schema
 * const teamSchema = schema({
 *   users: [userSchema],
 * })
 */
const schema = (...groups: (Object | Schema | SchemaGroup)[]): Schema =>
  groups.reduce((finalSchema, currentGroup) => {
    if (typeof currentGroup === 'function') {
      return currentGroup(finalSchema)
    } else if (isSchema(currentGroup)) {
      return currentGroup
    }
    return group(currentGroup)(finalSchema)
  }, defaultSchema())

export default schema
