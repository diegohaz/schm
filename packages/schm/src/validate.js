// @flow
import parse from './parse'
import { mapValuesToSchema } from './utils'
import type { Schema, ValidationError } from './types'

const isPrimitive = arg =>
  arg == null || (typeof arg !== 'object' && typeof arg !== 'function')

const replaceMessage = (
  message: string,
  paramName: string,
  value: any,
  validatorName: string
): string => (
  message
    .replace(/\{(PARAM|PATH)\}/g, paramName)
    .replace(/\{VALUE\}/g, value)
    .replace(/\{(VALIDATOR|TYPE)\}/g, validatorName)
)

const createErrorObject = (
  param: string,
  value: any,
  validator: string,
  option: any,
  message?: string
): ValidationError => ({
  param,
  ...isPrimitive(value) ? { value } : {},
  validator,
  ...isPrimitive(option) ? { [validator]: option } : {},
  ...message ? { message: replaceMessage(message, param, value, validator) } : {},
})

/**
 * Validates a schema based on given values.
 * @example
 * const userSchema = schema({
 *   name: {
 *     type: String,
 *     required: true,
 *   },
 *   age: {
 *     type: Number,
 *     min: [18, 'Too young'],
 *   }
 * })
 *
 * validate({ name: 'John', age: 17 }, userSchema)
 *   .then((parsedValues) => {
 *     console.log('Yaay!', parsedValues)
 *   })
 *   .catch((errors) => {
 *     console.log('Oops!', errors)
 *   })
 *
 * \/*
 * Output:
 * Oops! [{
 *   param: 'age',
 *   value: 17,
 *   validator: 'min',
 *   min: 18,
 *   message: 'Too young',
 * }]
 * *\/
 */
const validate = (values?: Object = {}, schema: Schema): Promise<ValidationError[]> => {
  const parsed = parse(values, schema)
  const promises = []
  const errors = []

  mapValuesToSchema(parsed, schema, (value, options, paramName, paramPath) => {
    let error

    Object.keys(options).forEach((optionName) => {
      if (error) return

      const option = options[optionName]
      const validator = schema.validators[optionName]

      if (typeof validator === 'function') {
        const { valid, message } = validator(value, option, parsed, options, schema.params)
        if (!valid) {
          error = createErrorObject(paramPath, value, optionName, option, message)
          errors.push(error)
        } else if (typeof valid.catch === 'function') {
          promises.push(valid.catch(() => (
            Promise.reject(createErrorObject(paramPath, valid, optionName, option, message))
          )))
        }
      } else if (validator) {
        throw new Error(`[schm] ${paramName} validator must be a function`)
      }
    })
  })

  if (errors.length) {
    return Promise.reject(errors)
  } else if (promises.length) {
    return Promise.all(promises).then(() => parsed)
  }
  return Promise.resolve(parsed)
}

export default validate
