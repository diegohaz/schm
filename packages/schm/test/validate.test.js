import validatejs from "validate.js";
import schema from "../src/schema";
import validate from "../src/validate";

describe("validate", () => {
  it("fails with function", async () => {
    const schm = schema({
      foo: {
        type: String,
        validate: () => false
      }
    });
    await expect(validate({ foo: "wrong" }, schm)).rejects.toMatchSnapshot();
  });

  it("fails with promise", async () => {
    const schm = schema({
      foo: {
        type: String,
        validate: () => Promise.reject()
      }
    });
    await expect(validate({ foo: "wrong" }, schm)).rejects.toMatchSnapshot();
  });

  it("fails with function and message array", async () => {
    const schm = schema({
      foo: {
        type: String,
        validate: [() => false, "{VALUE} is wrong!"]
      }
    });
    await expect(validate({ foo: "wrong" }, schm)).rejects.toMatchSnapshot();
  });

  it("fails with function and message object", async () => {
    const schm = schema({
      foo: {
        type: String,
        validate: {
          validator: () => false,
          message: "{VALUE} is wrong!"
        }
      }
    });
    await expect(validate({ foo: "wrong" }, schm)).rejects.toMatchSnapshot();
  });

  it("fails with promise and message array", async () => {
    const schm = schema({
      foo: {
        type: String,
        validate: [() => Promise.reject(), "{VALUE} is wrong!"]
      }
    });
    await expect(validate({ foo: "wrong" }, schm)).rejects.toMatchSnapshot();
  });

  it("fails with promise and message object", async () => {
    const schm = schema({
      foo: {
        type: String,
        validate: {
          validator: () => Promise.reject(),
          message: "{VALUE} is wrong!"
        }
      }
    });
    await expect(validate({ foo: "wrong" }, schm)).rejects.toMatchSnapshot();
  });

  it("fails with nested param", async () => {
    const schm = schema({
      foo: {
        bar: [
          {
            type: String,
            validate: v => v !== "wrong"
          }
        ]
      }
    });
    await expect(
      validate({ foo: { bar: ["right", "wrong"] } }, schm)
    ).rejects.toMatchSnapshot();
  });

  it("fails with multiple function validators", async () => {
    const schm = schema({
      foo: {
        type: String,
        validate: [v => v !== "wrong", v => v !== "notright"]
      }
    });
    await expect(validate({ foo: "wrong" }, schm)).rejects.toMatchSnapshot();
    await expect(validate({ foo: "notright" }, schm)).rejects.toMatchSnapshot();
  });

  it("fails with multiple array validators", async () => {
    const schm = schema({
      foo: {
        type: String,
        validate: [
          [v => v !== "wrong", "{VALUE} is wrong!"],
          [v => v !== "notright", "{VALUE} is notright!"]
        ]
      }
    });
    await expect(validate({ foo: "wrong" }, schm)).rejects.toMatchSnapshot();
    await expect(validate({ foo: "notright" }, schm)).rejects.toMatchSnapshot();
  });

  it("fails with multiple object validators", async () => {
    const schm = schema({
      foo: {
        type: String,
        validate: [
          { validator: v => v !== "wrong", message: "{VALUE} is wrong!" },
          { validator: v => v !== "notright", message: "{VALUE} is notright!" }
        ]
      }
    });
    await expect(validate({ foo: "wrong" }, schm)).rejects.toMatchSnapshot();
    await expect(validate({ foo: "notright" }, schm)).rejects.toMatchSnapshot();
  });

  it("fails with multiple params", async () => {
    const schm = schema({
      foo: { type: String, validate: () => false },
      bar: { type: String, validate: () => false }
    });
    await expect(
      validate({ foo: "wrong", bar: "wrong" }, schm)
    ).rejects.toMatchSnapshot();
  });

  it("passes with function", async () => {
    const schm = schema({
      foo: {
        type: String,
        validate: () => true
      }
    });
    await expect(validate({ foo: "right" }, schm)).resolves.toEqual({
      foo: "right"
    });
  });

  it("passes with promise", async () => {
    const schm = schema({
      foo: {
        type: String,
        validate: () => Promise.resolve()
      }
    });
    await expect(validate({ foo: "right" }, schm)).resolves.toEqual({
      foo: "right"
    });
  });

  it("throws if value is not a function", () => {
    const schm = schema({
      foo: {
        type: String,
        validate: "foo"
      }
    });
    expect(() => validate({ foo: "right" }, schm)).toThrow();
  });
});

describe("type", () => {
  it("fails with nested schema validator", async () => {
    const schm = schema({
      foo: {
        type: String,
        validate: [() => false, "foo"]
      }
    });
    const schm2 = schema({
      foos: [schm]
    });
    const values = { foos: [{ foo: "bar" }, { foo: "baz" }] };
    await expect(validate(values, schm2)).rejects.toMatchSnapshot();
  });

  it("calls custom nested schema validator", async () => {
    const customValidator = constraints => previous =>
      previous.merge({
        validate(values, paramPathPrefix) {
          const parsed = previous.parse(values);
          const transformKeys = object =>
            Object.keys(object).reduce(
              (finalObject, key) => ({
                ...finalObject,
                [[paramPathPrefix, key].join(".")]: object[key]
              }),
              {}
            );
          const errors = validatejs(values, constraints);
          return errors
            ? Promise.reject(transformKeys(errors))
            : Promise.resolve(parsed);
        }
      });
    const schm = schema(
      { foo: String },
      customValidator({ foo: { presence: true } })
    );
    const schm2 = schema({
      foos: [schm],
      bar: { type: String, required: true }
    });
    const values = { foos: [{}, { foo: "baz" }] };
    await expect(validate(values, schm2)).rejects.toMatchSnapshot();
  });
});

describe("required", () => {
  it("fails with value", async () => {
    const schm = schema({
      foo: {
        type: String,
        required: true
      }
    });
    await expect(validate(undefined, schm)).rejects.toMatchSnapshot();
  });

  it("fails with value and message array", async () => {
    const schm = schema({
      foo: {
        type: String,
        required: [true, "foo"]
      }
    });
    await expect(validate(undefined, schm)).rejects.toMatchSnapshot();
  });

  it("passes", async () => {
    const schm = schema({
      foo: {
        type: String,
        required: true
      }
    });
    await expect(validate({ foo: "foo" }, schm)).resolves.toEqual({
      foo: "foo"
    });
  });

  it("passes when option value is falsy", async () => {
    const schm = schema({
      foo: {
        type: String,
        required: false
      }
    });
    await expect(validate(undefined, schm)).resolves.toEqual({
      foo: undefined
    });
  });
});

describe("match", () => {
  it("fails with value", async () => {
    const schm = schema({
      foo: {
        type: String,
        match: /foo/
      }
    });
    await expect(validate({ foo: "bar" }, schm)).rejects.toMatchSnapshot();
  });

  it("fails with value and message array", async () => {
    const schm = schema({
      foo: {
        type: String,
        match: [/foo/, "foo"]
      }
    });
    await expect(validate({ foo: "bar" }, schm)).rejects.toMatchSnapshot();
  });

  it("passes", async () => {
    const schm = schema({
      foo: {
        type: String,
        match: /foo/
      }
    });
    await expect(validate({ foo: "foo" }, schm)).resolves.toEqual({
      foo: "foo"
    });
  });

  it("throws if value is not a regex", () => {
    const schm = schema({
      foo: {
        type: String,
        match: "foo"
      }
    });
    expect(() => validate({ foo: "foo" }, schm)).toThrow();
  });
});

describe("enum", () => {
  it("fails with value", async () => {
    const schm = schema({
      foo: {
        type: String,
        enum: ["foo", "bar"]
      }
    });
    await expect(validate({ foo: "baz" }, schm)).rejects.toMatchSnapshot();
  });

  it("fails with value and message object", async () => {
    const schm = schema({
      foo: {
        type: String,
        enum: { values: ["foo", "bar"], message: "foo" }
      }
    });
    await expect(validate({ foo: "baz" }, schm)).rejects.toMatchSnapshot();
  });

  it("passes", async () => {
    const schm = schema({
      foo: {
        type: String,
        enum: ["foo", "bar"]
      }
    });
    await expect(validate({ foo: "foo" }, schm)).resolves.toEqual({
      foo: "foo"
    });
  });

  it("throws if value is not an array", () => {
    const schm = schema({
      foo: {
        type: String,
        enum: "foo"
      }
    });
    expect(() => validate({ foo: "foo" }, schm)).toThrow();
  });
});

describe("max", () => {
  it("fails with value", async () => {
    const schm = schema({
      foo: {
        type: Number,
        max: 1
      }
    });
    await expect(validate({ foo: 2 }, schm)).rejects.toMatchSnapshot();
  });

  it("fails with value and message array", async () => {
    const schm = schema({
      foo: {
        type: Number,
        max: [1, "foo"]
      }
    });
    await expect(validate({ foo: 2 }, schm)).rejects.toMatchSnapshot();
  });

  it("passes", async () => {
    const schm = schema({
      foo: {
        type: Number,
        max: 1
      }
    });
    await expect(validate({ foo: 1 }, schm)).resolves.toEqual({
      foo: 1
    });
  });
});

describe("min", () => {
  it("fails with value", async () => {
    const schm = schema({
      foo: {
        type: Number,
        min: 2
      }
    });
    await expect(validate({ foo: 1 }, schm)).rejects.toMatchSnapshot();
  });

  it("fails with value and message array", async () => {
    const schm = schema({
      foo: {
        type: Number,
        min: [2, "foo"]
      }
    });
    await expect(validate({ foo: 1 }, schm)).rejects.toMatchSnapshot();
  });

  it("passes", async () => {
    const schm = schema({
      foo: {
        type: Number,
        min: 2
      }
    });
    await expect(validate({ foo: 2 }, schm)).resolves.toMatchSnapshot();
  });
});

describe("maxlength", () => {
  it("fails with value", async () => {
    const schm = schema({
      foo: {
        type: String,
        maxlength: 1
      }
    });
    await expect(validate({ foo: "12" }, schm)).rejects.toMatchSnapshot();
  });

  it("fails with value and message array", async () => {
    const schm = schema({
      foo: {
        type: String,
        maxlength: [1, "foo"]
      }
    });
    await expect(validate({ foo: "12" }, schm)).rejects.toMatchSnapshot();
  });

  it("passes", async () => {
    const schm = schema({
      foo: {
        type: String,
        maxlength: 1
      }
    });
    await expect(validate({ foo: "1" }, schm)).resolves.toEqual({
      foo: "1"
    });
  });
});

describe("minlength", () => {
  it("fails with value", async () => {
    const schm = schema({
      foo: {
        type: String,
        minlength: 2
      }
    });
    await expect(validate({ foo: "1" }, schm)).rejects.toMatchSnapshot();
  });

  it("fails with value and message array", async () => {
    const schm = schema({
      foo: {
        type: String,
        minlength: [2, "foo"]
      }
    });
    await expect(validate({ foo: "1" }, schm)).rejects.toMatchSnapshot();
  });

  it("passes", async () => {
    const schm = schema({
      foo: {
        type: String,
        minlength: 2
      }
    });
    await expect(validate({ foo: "12" }, schm)).resolves.toEqual({
      foo: "12"
    });
  });
});

test("without schema", async () => {
  const params = { foo: { type: String, required: true } };
  await expect(validate({}, params)).rejects.toMatchSnapshot();
});

test("custom option", async () => {
  const schm = schema({
    foo: {
      type: String,
      bar: "baz",
      required: true
    }
  });
  await expect(validate({ foo: "12" }, schm)).resolves.toEqual({
    foo: "12"
  });
});

test("validator must be a function", () => {
  const schm = schema(
    {
      foo: {
        type: String,
        bar: "baz"
      }
    },
    previous =>
      previous.merge({
        validators: {
          bar: true
        }
      })
  );
  expect(() => validate({}, schm)).toThrow();
});
