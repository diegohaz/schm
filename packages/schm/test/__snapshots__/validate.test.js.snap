// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enum fails with value 1`] = `
Array [
  Object {
    "message": "foo must be one of the following: foo, bar",
    "param": "foo",
    "validator": "enum",
    "value": "baz",
  },
]
`;

exports[`enum fails with value and message object 1`] = `
Array [
  Object {
    "message": "foo",
    "param": "foo",
    "validator": "enum",
    "value": "baz",
  },
]
`;

exports[`match fails with value 1`] = `
Array [
  Object {
    "message": "foo does not match",
    "param": "foo",
    "validator": "match",
    "value": "bar",
  },
]
`;

exports[`match fails with value and message array 1`] = `
Array [
  Object {
    "message": "foo",
    "param": "foo",
    "validator": "match",
    "value": "bar",
  },
]
`;

exports[`max fails with value 1`] = `
Array [
  Object {
    "max": 1,
    "message": "foo must be lower than or equal 1",
    "param": "foo",
    "validator": "max",
    "value": 2,
  },
]
`;

exports[`max fails with value and message array 1`] = `
Array [
  Object {
    "max": 1,
    "message": "foo",
    "param": "foo",
    "validator": "max",
    "value": 2,
  },
]
`;

exports[`maxlength fails with value 1`] = `
Array [
  Object {
    "maxlength": 1,
    "message": "foo length must be lower than or equal 1",
    "param": "foo",
    "validator": "maxlength",
    "value": "12",
  },
]
`;

exports[`maxlength fails with value and message array 1`] = `
Array [
  Object {
    "maxlength": 1,
    "message": "foo",
    "param": "foo",
    "validator": "maxlength",
    "value": "12",
  },
]
`;

exports[`min fails with value 1`] = `
Array [
  Object {
    "message": "foo must be greater than or equal 2",
    "min": 2,
    "param": "foo",
    "validator": "min",
    "value": 1,
  },
]
`;

exports[`min fails with value and message array 1`] = `
Array [
  Object {
    "message": "foo",
    "min": 2,
    "param": "foo",
    "validator": "min",
    "value": 1,
  },
]
`;

exports[`min passes 1`] = `
Object {
  "foo": 2,
}
`;

exports[`minlength fails with value 1`] = `
Array [
  Object {
    "message": "foo length must be greater than or equal 2",
    "minlength": 2,
    "param": "foo",
    "validator": "minlength",
    "value": "1",
  },
]
`;

exports[`minlength fails with value and message array 1`] = `
Array [
  Object {
    "message": "foo",
    "minlength": 2,
    "param": "foo",
    "validator": "minlength",
    "value": "1",
  },
]
`;

exports[`required fails with value 1`] = `
Array [
  Object {
    "message": "foo is required",
    "param": "foo",
    "required": true,
    "validator": "required",
    "value": undefined,
  },
]
`;

exports[`required fails with value and message array 1`] = `
Array [
  Object {
    "message": "foo",
    "param": "foo",
    "required": true,
    "validator": "required",
    "value": undefined,
  },
]
`;

exports[`type calls custom nested schema validator 1`] = `
Array [
  Object {
    "message": "bar is required",
    "param": "bar",
    "required": true,
    "validator": "required",
    "value": undefined,
  },
  Object {
    "foos.0.foo": Array [
      "Foo can't be blank",
    ],
  },
]
`;

exports[`type fails with nested schema validator 1`] = `
Array [
  Object {
    "message": "foo",
    "param": "foos.0.foo",
    "validator": "validate",
    "value": "bar",
  },
  Object {
    "message": "foo",
    "param": "foos.1.foo",
    "validator": "validate",
    "value": "baz",
  },
]
`;

exports[`validate fails with function 1`] = `
Array [
  Object {
    "param": "foo",
    "validator": "validate",
    "value": "wrong",
  },
]
`;

exports[`validate fails with function and message array 1`] = `
Array [
  Object {
    "message": "wrong is wrong!",
    "param": "foo",
    "validator": "validate",
    "value": "wrong",
  },
]
`;

exports[`validate fails with function and message object 1`] = `
Array [
  Object {
    "message": "wrong is wrong!",
    "param": "foo",
    "validator": "validate",
    "value": "wrong",
  },
]
`;

exports[`validate fails with multiple array validators 1`] = `
Array [
  Object {
    "message": "wrong is wrong!",
    "param": "foo",
    "validator": "validate",
    "value": "wrong",
  },
]
`;

exports[`validate fails with multiple array validators 2`] = `
Array [
  Object {
    "message": "notright is notright!",
    "param": "foo",
    "validator": "validate",
    "value": "notright",
  },
]
`;

exports[`validate fails with multiple function validators 1`] = `
Array [
  Object {
    "param": "foo",
    "validator": "validate",
    "value": "wrong",
  },
]
`;

exports[`validate fails with multiple function validators 2`] = `
Array [
  Object {
    "param": "foo",
    "validator": "validate",
    "value": "notright",
  },
]
`;

exports[`validate fails with multiple object validators 1`] = `
Array [
  Object {
    "message": "wrong is wrong!",
    "param": "foo",
    "validator": "validate",
    "value": "wrong",
  },
]
`;

exports[`validate fails with multiple object validators 2`] = `
Array [
  Object {
    "message": "notright is notright!",
    "param": "foo",
    "validator": "validate",
    "value": "notright",
  },
]
`;

exports[`validate fails with multiple params 1`] = `
Array [
  Object {
    "param": "foo",
    "validator": "validate",
    "value": "wrong",
  },
  Object {
    "param": "bar",
    "validator": "validate",
    "value": "wrong",
  },
]
`;

exports[`validate fails with nested param 1`] = `
Array [
  Object {
    "param": "foo.bar.1",
    "validator": "validate",
    "value": "wrong",
  },
]
`;

exports[`validate fails with promise 1`] = `
Array [
  Object {
    "param": "foo",
    "validator": "validate",
    "value": "wrong",
  },
]
`;

exports[`validate fails with promise and message array 1`] = `
Array [
  Object {
    "message": "wrong is wrong!",
    "param": "foo",
    "validator": "validate",
    "value": "wrong",
  },
]
`;

exports[`validate fails with promise and message object 1`] = `
Array [
  Object {
    "message": "wrong is wrong!",
    "param": "foo",
    "validator": "validate",
    "value": "wrong",
  },
]
`;

exports[`without schema 1`] = `
Array [
  Object {
    "message": "foo is required",
    "param": "foo",
    "required": true,
    "validator": "required",
    "value": undefined,
  },
]
`;
